Content.Server/Radiation/Systems/RadiationSystem.GridCast.cs

```diff
@@ -0,0 +1,277 @@
+using System.Linq;
+using Content.Server.Radiation.Components;
+using Content.Shared.Physics;
+using Content.Shared.Radiation.Components;
+using Content.Shared.Radiation.Events;
+using Content.Shared.Radiation.Systems;
+using Robust.Shared.Map;
+using Robust.Shared.Physics;
+using Robust.Shared.Timing;
+
+namespace Content.Server.Radiation.Systems;
+
+// main algorithm that fire radiation rays to target
+public partial class RadiationSystem
+{
+    private void UpdateGridcast()
+    {
+        // should we save debug information into rays?
+        // if there is no debug sessions connected - just ignore it
+        var saveVisitedTiles = _debugSessions.Count > 0;
+
+        var stopwatch = new Stopwatch();
+        stopwatch.Start();
+
+        var sources = EntityQuery<RadiationSourceComponent, TransformComponent>().ToArray();
+        var destinations = EntityQuery<RadiationReceiverComponent, TransformComponent>().ToArray();
+        var blockerQuery = GetEntityQuery<RadiationBlockerComponent>();
+        var resistanceQuery = GetEntityQuery<RadiationGridResistanceComponent>();
+        var transformQuery = GetEntityQuery<TransformComponent>();
+
+        // trace all rays from rad source to rad receivers
+        var rays = new List<RadiationRay>();
+        var receivedRads = new List<(RadiationReceiverComponent, float)>();
+        foreach (var (dest, destTrs) in destinations)
+        {
+            var destWorld = _transform.GetWorldPosition(destTrs, transformQuery);
+
+            var rads = 0f;
+            foreach (var (source, sourceTrs) in sources)
+            {
+                // send ray towards destination entity
+                var ray = Irradiate(sourceTrs.Owner, sourceTrs, destTrs.Owner, destTrs, destWorld,
+                    source.Intensity, source.Slope, saveVisitedTiles,
+                    blockerQuery, resistanceQuery, transformQuery);
+                if (ray == null)
+                    continue;
+
+                // save ray for debug
+                rays.Add(ray);
+
+                // add rads to total rad exposure
+                if (ray.ReachedDestination)
+                    rads += ray.Rads;
+            }
+
+            receivedRads.Add((dest, rads));
+        }
+
+        // update information for debug overlay
+        var elapsedTime = stopwatch.Elapsed.TotalMilliseconds;
+        var totalSources = sources.Length;
+        var totalReceivers = destinations.Length;
+        var ev = new OnRadiationOverlayUpdateEvent(elapsedTime, totalSources, totalReceivers, rays);
+        UpdateDebugOverlay(ev);
```

Comment:
You should create the event inside of the method and early-out if no one's subscribed.